openforcefield/typing/engines/smirnoff/parameters.py

```diff
@@ -2147,6 +2470,136 @@ def create_force(self, system, topology, **kwargs):
+                    match.reference_molecule.assign_fractional_bond_orders(
+                            toolkit_registry=toolkit_registry,
+                            use_conformers=match.reference_molecule.conformers)
+
+            # scale k based on the bondorder of the central bond
+            if self.fractional_bondorder_interpolation == 'linear':
+                # we only interpolate on k
+                k = self._linear_interpolate_k(k_bondorder, central_bond.fractional_bond_order)
+            else:
+                raise Exception(
+                    "Fractional bondorder treatment {} is not implemented.".
+                    format(self.fractional_bondorder_method))
+                
+            # add a torsion with given parameters for topology atoms
+            force.addTorsion(atom_indices[0], atom_indices[1],
+                             atom_indices[2], atom_indices[3], periodicity,
+                             phase, k/idivf)
+
+    @staticmethod
+    def _linear_interpolate_k(k_bondorder, fractional_bond_order):
+        
+        # pre-empt case where no interpolation is necessary
+        if fractional_bond_order in k_bondorder:
+            return k_bondorder[fractional_bond_order]
+
+        # TODO: error out for nonsensical fractional bond orders
+
+        # find the nearest bond_order beneath our fractional value
+        try:
+            below = max(bo for bo in k_bondorder if bo < fractional_bond_order)
+        except ValueError:
+            below = None
+
+        # find the nearest bond_order above our fractional value
+        try:
+            above = min(bo for bo in k_bondorder if bo > fractional_bond_order)
+        except ValueError:
+            above = None
+
+        # handle case where we can clearly interpolate
+        if (above is not None) and (below is not None):
+            return (k_bondorder[below]
+                    + (k_bondorder[above] - k_bondorder[below])
+                    * ((fractional_bond_order - below)/(above - below)))
+
+        # error if we can't hope to interpolate at all
+        elif (above is None) and (below is None):
+            raise NotImplementedError(
+                    "Failed to find interpolation references for partial "
+                    "bond order '{}'".format(fractional_bond_order))
+
+        # extrapolate for fractional bond orders below our lowest defined bond order
+        elif below is None:
+            bond_orders = sorted(k_bondorder)
+            k = (k_bondorder[bond_orders[0]]
+                    - ((k_bondorder[bond_orders[1]] - k_bondorder[bond_orders[0]])
+                       /(bond_orders[1] - bond_orders[0]))
+                    * (bond_orders[0] - fractional_bond_order))
+            if k/k.unit < 0:
+                raise ValueError(
+                        "Extrapolated k for fractional bond order "
+                        "'{}' below zero".format(fractional_bond_order))
```

Comment:
```suggestion
```

Reply:
(blocking until we have consensus)@jchodera @davidlmobley @leeping Any thoughts on what we should do if torsion interpolation/extrapolation yields a negative `k`? My instinct is to allow it, so I'm suggesting we not raise an error.